src/plugins/intel_gpu/tests/test_cases/gemm_gpu_test.cpp

```diff
@@ -342,6 +342,178 @@ TEST(gemm_gpu, dynamic) {
     }
 }
 
+TEST(gemm_gpu, dynamic_multi_inference_same_shape) {
+    auto& engine = get_test_engine();
+
+    auto in1_dyn_layout = layout{ ov::PartialShape{ 1, 1, ov::Dimension(1, 10), 4 }, data_types::f32, format::bfyx };
+    auto in1_actual_layout = layout{ ov::PartialShape{ 1, 1, 3, 4 }, data_types::f32, format::bfyx };
+    auto in2_layout = layout{ ov::PartialShape{ 4, 1 }, data_types::f32, format::bfyx };
+    auto input1_1 = engine.allocate_memory(in1_actual_layout);
+    auto input1_2 = engine.allocate_memory(in1_actual_layout);
+    auto input2 = engine.allocate_memory(in2_layout);
+
+    std::vector<float> input1_data1 = {
+        1.f, -2.f, 3.f, -4.f,
+        5.f, 6.f, 1.f, 2.f,
+        3.f, 3.f, 2.f, -1.f,
+    };
+    std::vector<float> input1_data2 = {
+        -1.f, 2.f, -3.f, 4.f,
+        5.f, 6.f, -1.f, 2.f,
+        3.f, -3.f, 2.f, 1.f,
+    };
+    std::vector<float> input2_data = {
+        2.f, 5.f, -4.f, -7.f,
+    };
+    set_values(input1_1, input1_data1);
+    set_values(input1_2, input1_data2);
+    set_values(input2, input2_data);
+
+    std::vector<float> out_data1 = {
+        8.f, 22.f, 20.f
+    };
+    std::vector<float> out_data2 = {
+        -8.f, 30.f, -24.f
+    };
+
+    topology topology;
+    topology.add(input_layout("input1", in1_dyn_layout),
+                 input_layout("input2", in2_layout),
+                 gemm("gemm", { input_info("input1"), input_info("input2") }, data_types::f32, false, false, 1.0f, 0.0f, 4, 2)
+    );
+
+    ExecutionConfig config;
+    config.set_property(ov::intel_gpu::optimize_data(true));
+    config.set_property(ov::intel_gpu::allow_new_shape_infer(true));
+    network network(engine, topology, config);
+
+    {
+        network.set_input_data("input1", input1_1);
+        network.set_input_data("input2", input2);
+
+        auto outputs = network.execute();
+        ASSERT_EQ(outputs.size(), size_t(1));
+        ASSERT_EQ(outputs.begin()->first, "gemm");
+
+        auto prog = network.get_program();
+        auto& node = prog->get_node("gemm");
+        auto kernel_name = node.get_selected_impl()->get_kernel_name();
+        ASSERT_EQ(kernel_name, "gemm_tiled_opt");
```

Comment:
I'd suggest changing this check to something like:
```
    auto impl = node.get_selected_impl();
    ASSERT_TRUE(impl != nullptr);
    ASSERT_TRUE(impl->is_dynamic());
```

Reply:
done